app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,555 @@
+  public static final String ACTION_START_SERVER = "start_server";
+  public static final String ACTION_STOP_SERVER = "stop_server";
+  public static final String SELECTED_ZIM_PATHS_KEY = "selected_zim_paths";
+  private static final String IP_STATE_KEY = "ip_state_key";
+  private static final String TAG = "ZimHostActivity";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  private static final int LOCATION_SETTINGS_PERMISSION_RESULT = 101;
+  Intent serviceIntent;
+  private Task<LocationSettingsResponse> task;
+  ProgressDialog progressDialog;
+
+  private BooksOnDiskAdapter booksAdapter;
+  BookOnDiskDelegate.BookDelegate bookDelegate;
+  HotspotService hotspotService;
+  private String ip;
+  private ServiceConnection serviceConnection;
+  ArrayList<String> selectedBooksPath = new ArrayList<>();
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    setUpToolbar();
+
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(IP_STATE_KEY);
+      layoutServerStarted();
+    }
+    bookDelegate =
+        new BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil,
+            null,
+            null,
+            bookOnDiskItem -> {
+              select(bookOnDiskItem);
+              return Unit.INSTANCE;
+            });
+    bookDelegate.setSelectionMode(SelectionMode.MULTI);
+    booksAdapter = new BooksOnDiskAdapter(bookDelegate,
+        BookOnDiskDelegate.LanguageDelegate.INSTANCE
+    );
+
+    presenter.attachView(this);
+
+    presenter.loadBooks();
+    recyclerViewZimHost.setAdapter(booksAdapter);
+
+    serviceConnection = new ServiceConnection() {
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        HotspotService.HotspotBinder binder = (HotspotService.HotspotBinder) service;
+        hotspotService = binder.getService();
+        hotspotService.registerCallBack(ZimHostActivity.this);
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName arg0) {
+      }
+    };
+
+    serviceIntent = new Intent(this, HotspotService.class);
+
+    startServerButton.setOnClickListener(v -> {
+      //Get the path of ZIMs user has selected
+      if (!ServerUtils.isServerStarted) {
+        getSelectedBooksPath();
+        if (selectedBooksPath.size() > 0) {
+          startHotspotHelper();
+        } else {
+          Toast.makeText(ZimHostActivity.this, R.string.no_books_selected_toast_message,
+              Toast.LENGTH_SHORT).show();
+        }
+      } else {
+        startHotspotHelper();
+      }
+    });
+  }
+
+  private void startHotspotHelper() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      toggleHotspot();
+    } else {
+      //TO DO: show Dialog() + within that add check mobile Data check later.
```

Comment:
is this actually a `TODO`? if so delete the space between the `to` & `do`, if it isn't spelled properly tooling won't pickt it up

Reply:
Fixed in 8300dea416556ec795630c5f22243c3d22087abc